<h1>Analyzing Used Car Listings on eBay Kleinanzeigen</h1>
We will be working on a dataset of used cars from eBay Kleinanzeigen, a classifieds section of the German eBay website.

The dataset was originally scraped and uploaded to Kaggle. The version of the dataset we are working with is a sample of 50,000 data points that was prepared by Dataquest including simulating a less-cleaned version of the data.

The aim of this project is to clean the data and analyze the included used car listings. You'll also become familiar with some of the unique benefits jupyter notebook provides for pandas.

Let's start by importing the libraries we need and reading the dataset into pandas.



In [1]:
import pandas as pd
import numpy as np

autos = pd.read_csv('autos.csv', encoding="Latin-1")

In [2]:
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       seller offerType    price   abtest vehicleType  yearOfRegistration  \
0      privat   Angebot   $5,000  control         bus                2004   
1      privat   Angebot   $8,500  control   limousine                1997   
2      privat   Angebot   $8,990     test   limousine                2009   
3      privat   Angebot   $4,350  control  kleinwagen                2007   
4      privat   Angebot   $1,350     test       kombi                2003   
...       ...       ...      ...      ...         ...                 ...   
49995  privat   Angebot  $24,900  control   limousine                2011   
49996  privat   Angebot   $1,980  control      cabrio                1996   
49997  privat   Angebot  $13,200     test      cabrio                2014   
49998  privat   Angebot  $22,900  control       kombi                2013   
49999  privat   Angebot   $1,250  control   limousine                1996   

         gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0        manuell      158  andere  150,000km                    3      lpg   
1      automatik      286     7er  150,000km                    6   benzin   
2        manuell      102    golf   70,000km                    7   benzin   
3      automatik       71  fortwo   70,000km                    6   benzin   
4        manuell        0   focus  150,000km                    7   benzin   
...          ...      ...     ...        ...                  ...      ...   
49995  automatik      239      q5  100,000km                    1   diesel   
49996    manuell       75   astra  150,000km                    5   benzin   
49997  automatik       69     500    5,000km                   11   benzin   
49998    manuell      150      a3   40,000km                   11   diesel   
49999    manuell      101  vectra  150,000km                    1   benzin   

            brand notRepairedDamage          dateCreated  nrOfPictures  \
0         peugeot              nein  2016-03-26 00:00:00             0   
1             bmw              nein  2016-04-04 00:00:00             0   
2      volkswagen              nein  2016-03-26 00:00:00             0   
3           smart              nein  2016-03-12 00:00:00             0   
4            ford              nein  2016-04-01 00:00:00             0   
...           ...               ...                  ...           ...   
49995        audi              nein  2016-03-27 00:00:00             0   
49996        opel              nein  2016-03-28 00:00:00             0   
49997        fiat              nein  2016-04-02 00:00:00             0   
49998        audi              nein  2016-03-08 00:00:00             0   
49999        opel              nein  2016-03-13 00:00:00             0   

       postalCode             lastSeen  
0           79588  2016-04-06 06:45:54  
1           71034  2016-04-06 14:45:08  
2           35394  2016-04-06 20:15:37  
3           33729  2016-03-15 03:16:28  
4           39218  2016-04-01 14:38:50  
...           ...                  ...  
4999

In [3]:
autos.info()
autos.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

Above the first few rows of the dataset are displayed, as well as some info on the columns and data types. There are several object types which indicate String types in Pandas. We also have int64 data types. It appears several columns have some null values indicated by the code in the non-Null Count column.

The column names will need to be modified to Snake Case format. Also we will modify some column names further to make them simpler and more readable.

In [4]:
cols = autos.columns.copy()
print(cols)

new_cols = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'ab_test', 'vehicle_type', 
           'registration_year', 'gearbox', 'power_PS', 'model', 'odometer', 'registration_month', 
           'fuel_type', 'brand', 'unrepaired_damage', 'ad_created', 'num_photos', 'postal_code', 
           'last_seen']

autos.columns = new_cols
autos.head()

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')


date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price  ab_test vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_PS   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  num_photos  postal_code  \
0     peugeot              nein  2016-03-26 00:00:00           0        79588   
1         bmw              nein  2016-04-04 00:00:00           0        71034   
2  volkswagen              nein  2016-03-26 00:00:00           0        35394   
3       smart              nein  2016-03-12 00:00:00           0        33729   
4        ford              nein  2016-04-01 00:00:00           0        39218   

             last_seen  
0  2016-04-06 06:45:54  
1  2016-04-06 14:45:08  
2  2016-04-06 20:15:37  
3  2016-03-15 03:16:28  
4  2016-04-01 14:38:50

<h2>Data Exploration and Data Cleaning</h2>
Let's look for text columns where most values are the same as this can often be dropped since it does not provide very useful info. Also, we will identify numeric data sotred as string and clean and convert that.

In [5]:
autos.describe(include='all') #all to get categorical as well as numeric data

date_crawled         name  seller offer_type  price ab_test  \
count                 50000        50000   50000      50000  50000   50000   
unique                48213        38754       2          2   2357       2   
top     2016-03-30 19:48:02  Ford_Fiesta  privat    Angebot     $0    test   
freq                      3           78   49999      49999   1421   25756   
mean                    NaN          NaN     NaN        NaN    NaN     NaN   
std                     NaN          NaN     NaN        NaN    NaN     NaN   
min                     NaN          NaN     NaN        NaN    NaN     NaN   
25%                     NaN          NaN     NaN        NaN    NaN     NaN   
50%                     NaN          NaN     NaN        NaN    NaN     NaN   
75%                     NaN          NaN     NaN        NaN    NaN     NaN   
max                     NaN          NaN     NaN        NaN    NaN     NaN   

       vehicle_type  registration_year  gearbox      power_PS  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  num_photos   postal_code            last_seen  
count                 50000     50000.0  50000.000000                50000  
unique                   76         NaN           NaN                39481  
top     2016-04-03 00:00:00         NaN           NaN  2016-04-07 06:17:27  
freq                   1946         NaN           NaN                    8  
mean                    NaN         0.0  50813.627300                  NaN  
std                     NaN         0.0  25779.747957                  NaN  
min                     NaN         0.0   1067.000000                  NaN  
25%                     NaN         0.0  30451.000000                  NaN  
50%                     NaN         0.0  49577.000000                  NaN  
75%                     NaN         0.0  71540.000000                  NaN  
max                     NaN         0.0  99998.000000                  NaN

The seller column stands out as one to drop because it only has 2 unique values out of 50,000! Same for offer_type and ab_test and gearbox, and unrepaired_damage. Actually, almost all of these are binary choices so we can probably leave them after inspection except possibly for offer_type.

Vehicle and gearbox are missing data in about 10% of the rows so we should investigate this. Same for model, fuel_type and unrepaired_damage.

The numeric columns which need to be cleaned and converted include: price, odometer and num_photos.

First the numeric columns:

In [6]:
print(autos["num_photos"].value_counts())
print(autos["seller"].value_counts())
autos["offer_type"].value_counts()

0    50000
Name: num_photos, dtype: int64
privat        49999
gewerblich        1
Name: seller, dtype: int64


Angebot    49999
Gesuch         1
Name: offer_type, dtype: int64

num_photos has all zeros so we can drop this one. Same for seller and offer_type which are both essentially a single value.

In [7]:
autos.drop(["num_photos", "seller", "offer_type"], axis=1)

date_crawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

         price  ab_test vehicle_type  registration_year    gearbox  power_PS  \
0       $5,000  control          bus               2004    manuell       158   
1       $8,500  control    limousine               1997  automatik       286   
2       $8,990     test    limousine               2009    manuell       102   
3       $4,350  control   kleinwagen               2007  automatik        71   
4       $1,350     test        kombi               2003    manuell         0   
...        ...      ...          ...                ...        ...       ...   
49995  $24,900  control    limousine               2011  automatik       239   
49996   $1,980  control       cabrio               1996    manuell        75   
49997  $13,200     test       cabrio               2014  automatik        69   
49998  $22,900  control        kombi               2013    manuell       150   
49999   $1,250  control    limousine               1996    manuell       101   

        model   odometer  registration_month fuel_type       brand  \
0      andere  150,000km                   3       lpg     peugeot   
1         7er  150,000km                   6    benzin         bmw   
2        golf   70,000km                   7    benzin  volkswagen   
3      fortwo   70,000km                   6    benzin       smart   
4       focus  150,000km                   7    benzin        ford   
...       ...        ...                 ...       ...         ...   
49995      q5  100,000km                   1    diesel        audi   
49996   astra  150,000km                   5    benzin        opel   
49997     500    5,000km                  11    benzin        fiat   
49998      a3   40,000km                  11    diesel        audi   
49999  vectra  150,000km                   1    benzin        opel   

      unrepaired_damage           ad_created  postal_code            last_seen  
0                  nein  2016-03-26 00:00:00        79588  2016-04-06 06:45:54  
1                  nein  2016-04-04 00:00:00        71034  2016-04-06 14:45:08  
2                  nein  2016-03-26 00:00:00        35394  2016-04-06 20:15:37  
3                  nein  2016-03-12 00:00:00        33729  2016-03-15 03:16:28  
4                  nein  2016-04-01 00:00:00        39218  2016-04-01 14:38:50  
...                 ...                  ...          ...                  ...  
49995              nein  2016-03-27 00:00:00        82131  2016-04-01 13:47:40  
49996              nein  2016-03-28 00:00:00        44807  2016-04-02 14:18:02  
49997              nein  2016-04-02 00:00:00        73430  2016-04-04 11:47:27  
49998              nein  2016-03-08 00:00:00        35683  2016-04-05 16:45:07  
49999              nein  2016-03-13 00:00:00        45897  2016-04-06 21:18:48  

[50000 rows x 17 columns]

In [8]:
autos["unrepaired_damage"].value_counts()

nein    35232
ja       4939
Name: unrepaired_damage, dtype: int64

There is some distribution among yes and no values in unrepaired_damage so we will retain this column

Now let's remove non-numeric characters from the price and odometer columns then convert to numeric and rename the column.

In [9]:
print(autos["price"].value_counts())
autos["price"] = autos["price"].str.replace("$", "").str.replace(",", "").astype(int)
autos["price"].head()

$0            1421
$500           781
$1,500         734
$2,500         643
$1,200         639
              ... 
$20,399          1
$15,298          1
$3,890,000       1
$3,080           1
$2,035           1
Name: price, Length: 2357, dtype: int64


0    5000
1    8500
2    8990
3    4350
4    1350
Name: price, dtype: int64

In [10]:
print(autos["odometer"].value_counts())
autos["odometer"] = (autos["odometer"]
                        .str.replace(",","")
                        .str.replace("km","")
                        .astype(int))
autos.rename({"odometer":"odometer_km"}, inplace=True, axis=1)
autos["odometer_km"].head()

150,000km    32424
125,000km     5170
100,000km     2169
90,000km      1757
80,000km      1436
70,000km      1230
60,000km      1164
50,000km      1027
5,000km        967
40,000km       819
30,000km       789
20,000km       784
10,000km       264
Name: odometer, dtype: int64


0    150000
1    150000
2     70000
3     70000
4    150000
Name: odometer_km, dtype: int64

<h2>Explore Price and Odometer</h2>

In [19]:
print(autos["price"].unique().shape)
print(autos["price"].describe())
print(autos["price"].value_counts().head(15))
autos["price"].value_counts().sort_index(ascending=False).head(15)

(2357,)
count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64
0       1421
500      781
1500     734
2500     643
1000     639
1200     639
600      531
800      498
3500     498
2000     460
999      434
750      433
900      420
650      419
850      410
Name: price, dtype: int64


99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
350000      1
345000      1
299000      1
295000      1
265000      1
Name: price, dtype: int64

The average price is about 9,840 but we have many zero prices and several north of 1,000,000. It's unlikely to have cars over 1,000,000 on an ebay site so let's remove those and the zeros.

In [45]:
price_bool = autos[(autos["price"] != 0) & (autos["price"] <= 350000)]
price_bool["price"].describe()

count     48565.000000
mean       5888.935591
std        9059.854754
min           1.000000
25%        1200.000000
50%        3000.000000
75%        7490.000000
max      350000.000000
Name: price, dtype: float64

Above we removed zero prices and abnormally large prices above 350,000. We can see the mean came down to 5899.

In [51]:
print(autos["odometer_km"].value_counts())
autos["odometer_km"].describe()

150000    32424
125000     5170
100000     2169
90000      1757
80000      1436
70000      1230
60000      1164
50000      1027
5000        967
40000       819
30000       789
20000       784
10000       264
Name: odometer_km, dtype: int64


count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

Above we can see that 150,000 has the most common mileage in the dataset and higher mileage cars dominate. The mean is 125,733 km. We do not see any outliers worth dropping here. 

<h2>Exploring Date Columns</h2>

There are 5 columns that should represent date values. Some of these columns were created by the crawler, some came from the website itself. We can differentiate by referring to the data dictionary:

- `date_crawled`: added by the crawler
- `last_seen`: added by the crawler
- `ad_created`: from the website
- `registration_month`: from the website
- `registration_year`: from the website

Right now, the `date_crawled`, `last_seen`, and `ad_created` columns are all identified as string values by pandas. Because these three columns are represented as strings, we need to convert the data into a numerical representation so we can understand it quantitatively. The other two columns are represented as numeric values, so we can use methods like `Series.describe()` to understand the distribution without any extra data processing.

In [53]:
autos[["date_crawled", "ad_created", "last_seen"]][0:5]

date_crawled           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

Above these 3 columns are formatted as full datetime values. Let's find the distribution of the dates in these columns.

In [61]:
#print(autos["date_crawled"].value_counts())
#print(autos["ad_created"].value_counts())
#print(autos["last_seen"].value_counts())

autos["date_crawled"].str[:10].value_counts(dropna=False).sort_index()


2016-03-05    1269
2016-03-06     697
2016-03-07    1798
2016-03-08    1665
2016-03-09    1661
2016-03-10    1606
2016-03-11    1624
2016-03-12    1839
2016-03-13     778
2016-03-14    1831
2016-03-15    1699
2016-03-16    1475
2016-03-17    1576
2016-03-18     653
2016-03-19    1745
2016-03-20    1891
2016-03-21    1876
2016-03-22    1647
2016-03-23    1619
2016-03-24    1455
2016-03-25    1587
2016-03-26    1624
2016-03-27    1552
2016-03-28    1742
2016-03-29    1709
2016-03-30    1681
2016-03-31    1596
2016-04-01    1690
2016-04-02    1770
2016-04-03    1934
2016-04-04    1826
2016-04-05     655
2016-04-06     159
2016-04-07      71
Name: date_crawled, dtype: int64

The `date_crawled` column ranges from 3/5/2016 through 4/7/2016 with a pretty steady distribution.

In [74]:
print(autos["ad_created"].str[:10].value_counts().sort_index())
print(autos["ad_created"].str[:10].value_counts().sort_values(ascending=False))
print(autos["ad_created"].str[:10].value_counts().sort_index().tail(25))

2015-06-11       1
2015-08-10       1
2015-09-09       1
2015-11-10       1
2015-12-05       1
              ... 
2016-04-03    1946
2016-04-04    1844
2016-04-05     592
2016-04-06     163
2016-04-07      64
Name: ad_created, Length: 76, dtype: int64
2016-04-03    1946
2016-03-20    1893
2016-03-21    1886
2016-04-04    1844
2016-03-12    1831
              ... 
2016-01-03       1
2016-01-22       1
2016-02-22       1
2016-02-16       1
2015-06-11       1
Name: ad_created, Length: 76, dtype: int64
2016-03-14    1761
2016-03-15    1687
2016-03-16    1500
2016-03-17    1560
2016-03-18     686
2016-03-19    1692
2016-03-20    1893
2016-03-21    1886
2016-03-22    1640
2016-03-23    1609
2016-03-24    1454
2016-03-25    1594
2016-03-26    1628
2016-03-27    1545
2016-03-28    1748
2016-03-29    1707
2016-03-30    1672
2016-03-31    1596
2016-04-01    1690
2016-04-02    1754
2016-04-03    1946
2016-04-04    1844
2016-04-05     592
2016-04-06     163
2016-04-07      64
Name: ad_created, dty

The `ad_created` column has a wider and non-uniform distribution of its values. And the majority of its values are compressed into just a few weeks.

In [76]:
print(autos["last_seen"].str[:10].value_counts().sort_index())
print(autos["last_seen"].str[:10].value_counts().sort_values(ascending=False))
print(autos["last_seen"].str[:10].value_counts().sort_index().tail(25))

2016-03-05       54
2016-03-06      221
2016-03-07      268
2016-03-08      380
2016-03-09      493
2016-03-10      538
2016-03-11      626
2016-03-12     1191
2016-03-13      449
2016-03-14      640
2016-03-15      794
2016-03-16      822
2016-03-17     1396
2016-03-18      371
2016-03-19      787
2016-03-20     1035
2016-03-21     1037
2016-03-22     1079
2016-03-23      929
2016-03-24      978
2016-03-25      960
2016-03-26      848
2016-03-27      801
2016-03-28     1043
2016-03-29     1117
2016-03-30     1242
2016-03-31     1192
2016-04-01     1155
2016-04-02     1245
2016-04-03     1268
2016-04-04     1231
2016-04-05     6214
2016-04-06    11050
2016-04-07     6546
Name: last_seen, dtype: int64
2016-04-06    11050
2016-04-07     6546
2016-04-05     6214
2016-03-17     1396
2016-04-03     1268
2016-04-02     1245
2016-03-30     1242
2016-04-04     1231
2016-03-31     1192
2016-03-12     1191
2016-04-01     1155
2016-03-29     1117
2016-03-22     1079
2016-03-28     1043
2016-03-21

The `last_seen` column has a narrower distribution confined to a couple of months but it is skewed to the right with a majority of values falling in the very early time frame of the index.

In [77]:
autos["registration_year"].describe()

count    50000.000000
mean      2005.073280
std        105.712813
min       1000.000000
25%       1999.000000
50%       2003.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

The `registration_year` column above has some outlier values with a max of 9999 which is not a valid year, and a min of 1000.

In [91]:
print(autos["registration_year"].sort_values().head(20))
print(autos["registration_year"].sort_values().tail(30))
autos[autos["registration_year"].between(1910, 2016)]["registration_year"].describe()

22316    1000
49283    1001
24511    1111
35238    1500
10556    1800
32585    1800
28693    1910
42181    1910
15898    1910
3679     1910
30781    1910
33295    1910
45157    1910
22659    1910
46213    1910
21416    1927
22101    1929
11246    1931
2573     1934
2221     1934
Name: registration_year, dtype: int64
13347    2018
49354    2018
28335    2018
20154    2018
42025    2018
802      2018
20142    2018
45940    2018
42022    2018
38342    2019
5763     2019
49185    2019
27578    2800
4549     4100
453      4500
42079    4800
22799    5000
4164     5000
49153    5000
24519    5000
27618    5911
8360     6200
25003    8888
49910    9000
13559    9000
6308     9996
8012     9999
14341    9999
33950    9999
38076    9999
Name: registration_year, dtype: int64


count    48028.00000
mean      2002.80351
std          7.31085
min       1910.00000
25%       1999.00000
50%       2003.00000
75%       2008.00000
max       2016.00000
Name: registration_year, dtype: float64

It would make sense, per above, to throw out years prior to 1910 and years after 2016.

One thing that stands out from the exploration we did in the last screen is that the registration_year column contains some odd values. The minimum value is 1000, before cars were invented. The maximum value is 9999, many years into the future.

Because a car can't be first registered after the listing was seen, any vehicle with a registration year above 2016 is definitely inaccurate. Determining the earliest valid year is more difficult. Realistically, it could be somewhere in the first few decades of the 1900s.

Once we remove the outliers as seen above we can see normal descriptive statistics of this column.

<h2>Exploring Price by Brand</h2>

In [94]:
autos["brand"].value_counts(normalize=True)

volkswagen        0.21374
opel              0.10922
bmw               0.10858
mercedes_benz     0.09468
audi              0.08566
ford              0.06958
renault           0.04808
peugeot           0.02912
fiat              0.02616
seat              0.01882
skoda             0.01572
mazda             0.01514
nissan            0.01508
citroen           0.01402
smart             0.01402
toyota            0.01234
sonstige_autos    0.01092
hyundai           0.00976
volvo             0.00914
mini              0.00848
mitsubishi        0.00812
honda             0.00798
kia               0.00712
alfa_romeo        0.00658
porsche           0.00588
suzuki            0.00586
chevrolet         0.00566
chrysler          0.00362
dacia             0.00258
daihatsu          0.00256
jeep              0.00220
subaru            0.00218
land_rover        0.00198
saab              0.00160
daewoo            0.00158
trabant           0.00156
jaguar            0.00154
rover             0.00138
lancia      

Volkswagen is by far the most popular brand with more than 21% of the values. We will aggregate the brands which each represent at least 5% of the total counts in the `brand` column.

In [99]:
brand_counts = autos["brand"].value_counts(normalize=True)
common_brands = brand_counts[brand_counts > 0.05].index
common_brands

Index(['volkswagen', 'opel', 'bmw', 'mercedes_benz', 'audi', 'ford'], dtype='object')

In [108]:
brands_mean_prices = {}
for brand in common_brands:
    brand_table = autos[autos["brand"] == brand]
    mean_price = brand_table["price"].mean()
    brands_mean_prices[brand] = mean_price
print(brands_mean_prices)

{'volkswagen': 6384.167399644428, 'opel': 5106.092657022524, 'bmw': 8252.918953766808, 'mercedes_benz': 29511.955428812842, 'audi': 8965.560354891431, 'ford': 7105.662546708824}


Above we see the mean prices for the common brands of cars listed on eBay. Mercedez is the highest by far with Opel being the lowest.

<h2>Exploring mileage using the Pandas Dataframe and Series Constructors</h2>

In [119]:
brands_mean_mileage = {}
for brand in common_brands:
    brand_table = autos[autos["brand"] == brand]
    mean_mileage = brand_table["odometer_km"].mean()
    brands_mean_mileage[brand] = mean_mileage
print(brands_mean_mileage)

#convert dicts to series objects
price_series = pd.Series(brands_mean_prices)
print(price_series)
mileage_series = pd.Series(brands_mean_mileage)
print(mileage_series)

#convert price series to dataframe
df = pd.DataFrame(price_series, columns=['mean_price'])
df

#Assign mileage series to dataframe
df['mean_mileage'] = mileage_series
df

{'volkswagen': 128955.27276129878, 'opel': 129298.66324848929, 'bmw': 132521.64302818198, 'mercedes_benz': 130886.14279678918, 'audi': 129643.9411627364, 'ford': 124131.93446392642}
volkswagen        6384.167400
opel              5106.092657
bmw               8252.918954
mercedes_benz    29511.955429
audi              8965.560355
ford              7105.662547
dtype: float64
volkswagen       128955.272761
opel             129298.663248
bmw              132521.643028
mercedes_benz    130886.142797
audi             129643.941163
ford             124131.934464
dtype: float64


mean_price   mean_mileage
volkswagen      6384.167400  128955.272761
opel            5106.092657  129298.663248
bmw             8252.918954  132521.643028
mercedes_benz  29511.955429  130886.142797
audi            8965.560355  129643.941163
ford            7105.662547  124131.934464

The new DataFrame is displayed above. The mean mileages are pretty uniformly distributed so the mileage does not appear to have much of an effect on the mean price differences among brands.